pkg/solana/chainwriter/lookups.go

```diff
@@ -0,0 +1,356 @@
+}
+
+// AccountLookup dynamically derives an account address from args using a specified location path.
+type AccountLookup struct {
+	Name       string
+	Location   string
+	IsSigner   bool
+	IsWritable bool
+}
+
+// PDALookups generates Program Derived Addresses (PDA) by combining a derived public key with one or more seeds.
+type PDALookups struct {
+	Name string
+	// The public key of the PDA to be combined with seeds. If there are multiple PublicKeys
+	// there will be multiple PDAs generated by combining each PublicKey with the seeds.
+	PublicKey Lookup
+	// Seeds to be derived from an additional lookup
+	Seeds      []Lookup
+	IsSigner   bool
+	IsWritable bool
+	// OPTIONAL: On-chain location and type of desired data from PDA (e.g. a sub-account of the data account)
+	InternalField InternalField
+}
+
+type InternalField struct {
+	Type     reflect.Type
+	Location string
+}
+
+type ValueLookup struct {
+	Location string
+}
+
+// LookupTables represents a list of lookup tables that are used to derive addresses for a program.
+type LookupTables struct {
+	DerivedLookupTables []DerivedLookupTable
+	StaticLookupTables  []string
+}
+
+// DerivedLookupTable represents a lookup table that is used to derive addresses for a program.
+type DerivedLookupTable struct {
+	Name     string
+	Accounts Lookup
+}
+
+// AccountsFromLookupTable extracts accounts from a lookup table that was previously read and stored in memory.
+type AccountsFromLookupTable struct {
+	LookupTableName string
+	IncludeIndexes  []int
+}
+
+func (ac AccountConstant) Resolve(_ context.Context, _ any, _ map[string]map[string][]*solana.AccountMeta, _ client.Reader) ([]*solana.AccountMeta, error) {
+	address, err := solana.PublicKeyFromBase58(ac.Address)
+	if err != nil {
+		return nil, fmt.Errorf("error getting account from constant: %w", err)
+	}
+	return []*solana.AccountMeta{
+		{
+			PublicKey:  address,
+			IsSigner:   ac.IsSigner,
+			IsWritable: ac.IsWritable,
+		},
+	}, nil
+}
+
+func (al AccountLookup) Resolve(_ context.Context, args any, _ map[string]map[string][]*solana.AccountMeta, _ client.Reader) ([]*solana.AccountMeta, error) {
+	derivedValues, err := GetValuesAtLocation(args, al.Location)
+	if err != nil {
+		return nil, fmt.Errorf("error getting account from lookup: %w", err)
+	}
+
+	var metas []*solana.AccountMeta
+	for _, address := range derivedValues {
+		metas = append(metas, &solana.AccountMeta{
+			PublicKey:  solana.PublicKeyFromBytes(address),
+			IsSigner:   al.IsSigner,
+			IsWritable: al.IsWritable,
+		})
+	}
+	return metas, nil
+}
+
+func (alt AccountsFromLookupTable) Resolve(_ context.Context, _ any, derivedTableMap map[string]map[string][]*solana.AccountMeta, _ client.Reader) ([]*solana.AccountMeta, error) {
+	// Fetch the inner map for the specified lookup table name
+	innerMap, ok := derivedTableMap[alt.LookupTableName]
+	if !ok {
+		return nil, fmt.Errorf("lookup table not found: %s", alt.LookupTableName)
+	}
+
+	var result []*solana.AccountMeta
```

Comment:
nit: think we can initialize this with `len(innerMap)` to avoid reallocations
```suggestion
	result := make([]*solana.AccountMeta, 0, len(innerMap))
```

Reply:
Actually never mind, realized this isn't one to one